cpp/include/cudf/scalar/scalar.hpp

```diff
@@ -0,0 +1,161 @@
+   * @brief Construct a new scalar object by deep copying the contents of
+   *`other`.
+   *
+   * Uses the specified `stream` and device_memory_resource for all allocations
+   * and copies.
+   *
+   * @param other The `scalar` to copy
+   * @param stream The stream on which to execute all allocations and copies
+   * @param mr The resource to use for all allocations
+   *---------------------------------------------------------------------------**/
+  scalar(scalar const& other, cudaStream_t stream,
+         rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource());
+
+  /**---------------------------------------------------------------------------*
+   * @brief Move the contents from `other` to create a new scalar.
+   *
+   * After the move, `other.type() = {EMPTY}`
+   *
+   * @param other The scalar whose contents will be moved into the new scalar
+   *---------------------------------------------------------------------------**/
+  scalar(scalar&& other);
+
+  /**---------------------------------------------------------------------------*
+   * @brief Construct a new scalar from existing device memory.
+   *
+   * @note This constructor is primarily intended for use in scalar factory
+   * functions. 
+   *
+   * @param[in] dtype The element type
+   * @param[in] size The number of elements in the scalar
+   * @param[in] data The scalar's data
+   * @param[in] is_valid Optional, scalar's null value indicator bitmask. May
+   * be empty if `null_count` is 0 or `UNKNOWN_NULL_COUNT`.
+   *---------------------------------------------------------------------------**/
+  template <typename B1, typename B2 = rmm::device_scalar<bool>>
+  scalar(data_type dtype, size_type size, B1&& data, B2&& is_valid = {})
+      : _type{dtype},
+        _size{size},
+        _data{std::forward<B1>(data)},
+        _is_valid{std::forward<B2>(is_valid)} {}
+
+  /**---------------------------------------------------------------------------*
+   * @brief Returns the scalar's logical element type
+   *---------------------------------------------------------------------------**/
+  data_type type() const noexcept { return _type; }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Sets this scalar to null
+   *---------------------------------------------------------------------------**/
+  void set_null();
+
+  /**---------------------------------------------------------------------------*
+   * @brief Indicates whether the scalar contains a valid value
+   *
+   * @note Using the value when `is_valid() == false` is undefined behaviour
+   * 
+   * @return true Value is valid
+   * @return false Value is invalid
+   *---------------------------------------------------------------------------**/
+  bool is_valid() const;
+
+ protected:
+  data_type _type{EMPTY};      ///< Logical type of elements in the scalar
+  rmm::device_scalar<bool> _is_valid{};  ///< Device bool signifying validity
```

Comment:
What's the thought on making this live in device memory?

I'd expect most of the usage of `is_valid` would be from the host, e.g.,
```
if(scalar.is_valid){
   kernel<true><<<...>>>(...);
} else{
   kernel<false><<<...>>>(...);
}
```

Therefore I'd think it should live on the host. If it _does_ need to be passed into a kernel, it can just be passed by value:
```
__global__ void kernel(..., bool is_valid);
...

kernel<<<...>>>(..., scalar.is_valid());
```

Reply:
For returning. Scalars are not just for passing to device but also for getting values from them e.g. in series quantiles and reductions. Then we want the device to be able to write to some allocated memory. Yes, it’s easier for the null bool to simply be cudaMemCpy’d because it’ll always be a singular value, unlike string_view. But since we already have the value in the device, we have lost the convenience of forgoing a separate device view for the scalar. So we might as well put this in the device too.

*TLDR*: No benefit now in keeping it on host, only a disadvantage.

Reply:
Hm, I see your point about returning from a kernel. However, I don't think there's _no_ advantage to keeping it on the host. The disadvantage in keeping it in device memory is that accessing it on the host (which will be common) will incur a memcpy.

Reply:
Memoization to the rescue? Like you did for null  count. And the reverse, only create the device memory and copy it if anyone creates a device view for this.

Reply:
Is memoization future work or part of this PR?